src/main/kotlin/com/ampnet/headlesscmsservice/service/impl/MailServiceImpl.kt

```diff
@@ -0,0 +1,117 @@
+            (type != null && lang == null) -> getAllMailsByType(customMails, coop, type)
+            else -> getAllMailsByTypeAndLanguage(customMails, coop)
+        }
+    }
+
+    override fun updateMail(request: MailUpdateServiceRequest): MailResponse {
+        val mail = mailRepository.findByCoopAndOptionalTypeAndOptionalLang(
+            request.coop, request.type, request.lang
+        ).firstOrNull() ?: throw ResourceNotFoundException(
+            ErrorCode.CMS_MAIL_MISSING,
+            "${request.type} in ${request.lang} for coop: ${request.coop} is missing"
+        )
+        validateMailUpdateHasRequiredFields(request.content, request.type.getRequiredFields())
+        mail.title = request.title
+        mail.content = request.content
+        return MailResponse(mail)
+    }
+
+    private fun validateMailUpdateHasRequiredFields(mailContent: String, requiredFields: List<MailFieldName>) {
+        val missingFields = requiredFields.filterNot { mailContent.contains(it.value, false) }
+        if (missingFields.isNotEmpty()) throw InvalidRequestException(
+            ErrorCode.CMS_REQUIRED_FIELD_MISSING,
+            "Required fields are missing: ${ missingFields.joinToString() }"
+        )
+    }
+
+    private fun generateDefaultMail(
+        coop: String,
+        typeAndLang: MailTypeAndLang
+    ): MailResponse {
+        val content = translationService.getTranslation(typeAndLang.mailType.defaultTemplateKey, typeAndLang.lang)
+        val title = translationService.getTranslation(typeAndLang.mailType.defaultTitleKey, typeAndLang.lang)
+        return MailResponse(
+            UUID.randomUUID(), coop, title, content, typeAndLang.mailType,
+            typeAndLang.mailType.getRequiredFields().map { it.value }, typeAndLang.lang
+        )
+    }
+
+    private fun getAllMailsByLanguage(
+        customMails: List<MailResponse>,
+        coop: String,
+        lang: Lang
+    ): MailListResponse {
+        val mailByTypes = customMails.associateBy { it.type }.toMutableMap()
+        MailType.values().forEach { mailType ->
+            if (mailByTypes.containsKey(mailType).not()) {
+                mailByTypes[mailType] = generateDefaultMail(coop, MailTypeAndLang(mailType, lang))
+            }
+        }
+        return MailListResponse(mailByTypes.values.toList())
+    }
+
+    private fun getAllMailsByType(
+        customMails: List<MailResponse>,
+        coop: String,
+        mailType: MailType
+    ): MailListResponse {
+        val mailByLanguages = customMails.associateBy { it.lang }.toMutableMap()
+        Lang.values().forEach { lang ->
+            if (mailByLanguages.containsKey(lang).not()) {
+                mailByLanguages[lang] = generateDefaultMail(coop, MailTypeAndLang(mailType, lang))
+            }
+        }
+        return MailListResponse(mailByLanguages.values.toList())
+    }
+
+    private fun getAllMailsByTypeAndLanguage(
```

Comment:
This function returns all mails, not checking type and language.
```suggestion
    private fun getAllMails(
```